File: src/test/java/io/lsdconsulting/lsd/distributed/diagram/data/InteractionDataGeneratorShould.java
package io.lsdconsulting.lsd.distributed.diagram.data;

import io.lsdconsulting.lsd.distributed.model.InterceptedInteraction;
import org.junit.jupiter.api.Test;

import java.util.Collection;
import java.util.List;
import java.util.Map;

import static io.lsdconsulting.lsd.distributed.model.Type.*;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;

class InteractionDataGeneratorShould {

    private final InteractionDataGenerator underTest = new InteractionDataGenerator();

    @Test
    void buildWithRequestHeadersOnly() {
        Map<String, Collection<String>> requestHeaders = Map.of("name", List.of("value"));
        InterceptedInteraction interaction = InterceptedInteraction.builder()
                .type(REQUEST)
                .requestHeaders(requestHeaders)
                .responseHeaders(Map.of("name1", List.of("value2")))
                .body("someBody")
                .build();

        InteractionDataGenerator.InteractionData result = underTest.buildFrom(interaction);

        assertThat(result.getRequestHeaders(), is(requestHeaders));
        assertThat(result.getResponseHeaders(), is(nullValue()));
        assertThat(result.getHeaders(), is(nullValue()));
        assertThat(result.getBody(), is("someBody"));
    }

    @Test
    void buildWithResponseHeadersOnly() {
        Map<String, Collection<String>> responseHeaders = Map.of("name", List.of("value"));
        InterceptedInteraction interaction = InterceptedInteraction.builder()
                .type(RESPONSE)
                .requestHeaders(Map.of("name1", List.of("value2")))
                .responseHeaders(responseHeaders)
                .body("someBody")
                .build();

        InteractionDataGenerator.InteractionData result = underTest.buildFrom(interaction);

        assertThat(result.getRequestHeaders(), is(nullValue()));
        assertThat(result.getResponseHeaders(), is(responseHeaders));
        assertThat(result.getHeaders(), is(nullValue()));
        assertThat(result.getBody(), is("someBody"));
    }

    @Test
    void buildWithHeadersOnly() {
        Map<String, Collection<String>> headers = Map.of("name", List.of("value"));
        InterceptedInteraction interaction = InterceptedInteraction.builder()
                .type(PUBLISH)
                .requestHeaders(headers)
                .responseHeaders(Map.of("name1", List.of("value2")))
                .body("someBody")
                .build();

        InteractionDataGenerator.InteractionData result = underTest.buildFrom(interaction);

        assertThat(result.getRequestHeaders(), is(nullValue()));
        assertThat(result.getResponseHeaders(), is(nullValue()));
        assertThat(result.getHeaders(), is(headers));
        assertThat(result.getBody(), is("someBody"));
    }
}

File: src/test/kotlin/io/lsdconsulting/lsd/distributed/generator/diagram/data/InteractionDataGeneratorShould.kt
package io.lsdconsulting.lsd.distributed.generator.diagram.data

import io.lsdconsulting.lsd.distributed.access.model.InteractionType
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers
import org.hamcrest.Matchers.`is`
import org.junit.jupiter.api.Test
import java.time.ZoneId
import java.time.ZonedDateTime

internal class InteractionDataGeneratorShould {

    @Test
    fun `build with request headers only`() {
        val requestHeaders = mutableMapOf<String, Collection<String>>("name" to listOf("value"))
        val interaction = InterceptedInteraction(
            traceId = "traceId",
            interactionType = InteractionType.REQUEST,
            requestHeaders = requestHeaders,
            responseHeaders = mapOf("name1" to listOf("value2")),
            body = "someBody",
            serviceName = "serviceName",
            target = "target",
            path = "path",
            elapsedTime = 0,
            createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
        )
        val result = buildDataFrom(interaction)
        assertThat(result.requestHeaders, `is`(requestHeaders))
        assertThat(result.responseHeaders, `is`(Matchers.nullValue()))
        assertThat(result.headers, `is`(Matchers.nullValue()))
        assertThat(result.body, `is`("someBody"))
    }

    @Test
    fun `build with response headers only`() {
        val responseHeaders = mutableMapOf<String, Collection<String>>("name" to listOf("value"))
        val interaction = InterceptedInteraction(
            traceId = "traceId",
            interactionType = InteractionType.RESPONSE,
            requestHeaders = mutableMapOf("name1" to listOf("value2")),
            responseHeaders = responseHeaders,
            body = "someBody",
            serviceName = "serviceName",
            target = "target",
            path = "path",
            elapsedTime = 0,
            createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
        )
        val result = buildDataFrom(interaction)
        assertThat(result.requestHeaders, `is`(Matchers.nullValue()))
        assertThat(result.responseHeaders, `is`(responseHeaders))
        assertThat(result.headers, `is`(Matchers.nullValue()))
        assertThat(result.body, `is`("someBody"))
    }

    @Test
    fun `build with headers only`() {
        val headers = mutableMapOf<String, Collection<String>>("name" to listOf("value"))
        val interaction = InterceptedInteraction(
            traceId = "traceId",
            interactionType = InteractionType.PUBLISH,
            requestHeaders = headers,
            responseHeaders = mapOf("name1" to listOf("value2")),
            body = "someBody",
            serviceName = "serviceName",
            target = "target",
            path = "path",
            elapsedTime = 0,
            createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
        )
        val result = buildDataFrom(interaction)
        assertThat(result.requestHeaders, `is`(Matchers.nullValue()))
        assertThat(result.responseHeaders, `is`(Matchers.nullValue()))
        assertThat(result.headers, `is`(headers))
        assertThat(result.body, `is`("someBody"))
    }
}
